### PR TITLE
WIP: Added Thickness support for BorderThickness and CornerRadius

### PR DESCRIPTION
**Pull request by Gillibald**
I added support for BorderThickness and CornerRadius in a way it is available for WPF
ContentPresenter and Border use the same implementation to render.

This is unpolished code and there are no tests

### CODE REVIEW

**Comment by danwalmsley** [comment]
Will have a look shortly

**Comment by danwalmsley** [comment]
Ok iv spent quite a while testing this in the designer and seems to work for me, I need to spend more time looking at cornet radius though I only briefly tried this, but looks good so far.

**Comment by Gillibald** [comment]
ContentPresenter needs some extra test especially ArrangeOverrideImpl. I am not sure if all calculations are still valid. How can i make sure that everything works as intended?

**Comment by danwalmsley** [comment]
Normally compare behavior against WPF see how that behaves, check its consistent and add some tests to enforce that in future.

**Comment by Gillibald** [comment]
After i tested the rendering i realized that my geometry isnt rendered with sharp edges. Am i doing anything wrong setting up the DrawGeometry call? This needs to be improved to match visual quality of unified borders. A line from point (0,0) to (150,0) should have solid edges.

**Comment by grokys** [comment]
@Gillibald does a basic border render with sharp edges? Previously basic borders did, you should be preserving that behavior. You might want to check the rendering calls made on `master` against your rendering calls to make sure the same thing is getting drawn.

**Comment by Gillibald** [comment]
Basic rendering (uniform Thickness) has sharp edges as before. Just the complex rendering with a StreamGeometry doesnt produce the same quality. Even without calling ArcTo on the context i get rounded corners and the edges have some sub pixels.

The first image uses the old rendering with uniform borders. The second and third use complex rendering. No rounded shapes are created in the geometry still you get rounded corners and edges are not clean too.
borders

File: src/Avalonia.Visuals/CornerRadius.cs [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+        public static CornerRadius Parse(string s, CultureInfo culture)
+        {
+            var parts = s.Split(new[] { ',', ' ' }, StringSplitOptions.RemoveEmptyEntries)
+                .Select(x => x.Trim())
+                .ToList();
+
+            switch (parts.Count)
+            {
+                case 1:
```
[the remark this answers is not in the document]
Comment (Gillibald) [reply src/Avalonia.Visuals/CornerRadius.cs]: I have added that Feature. First thought it would not make sense but maybe it is more consistent.

**Comment by grokys** [comment]
@Gillibald this appears to be because you're not snapping geometry to pixel boundaries correctly. How familiar are you with 2D rendering techniques? This [stackoverflow question](https://stackoverflow.com/questions/10763762/pixels-in-direct2d) gives a brief introduction.

**Comment by Gillibald** [comment]
Thanks for the response i will have a look at it

**Comment by jkoritzinsky** [comment]
Also, we should revert back to the old behavior for uniform edge widths when possible since it is more efficient to draw rectangles than paths.

**Comment by Gillibald** [comment]
A fallback to old rendering is in place. I check for a uniform border and cornerradius and use the old rendering path then. After i changed my rendering code to take pixel boundaries into account i get a better image.

**Comment by Gillibald** [comment]
After more testing i realised that i had to pass null for pen in my DrawGeometry call. Now everything looks nice and clean. To allow a null value for pen I changed GeometryImpl.GetRenderBounds(Avalonia.Media.Pen pen) to accept a null value.

Do we need render tests? I think this PR is ready for review. I am open for improvements.

**Comment by Gillibald** [comment]
Looks like DrawRectangle isnt working properly if you set a bigger thickness like 20 you dont get a nice and clean filled area. The complex render code handles this fine and everything looks clean. 

1. Fast render Code (Fill./DrawRectangle)
2. Complex render Code (StreamGeometry)

borders

**Comment by grokys** [comment]
Ok, the problem isn't that `DrawRectangle` is incorrect - the problem is that the border is being drawn with transparent colors and the background is being drawn overlapping the border. You can see this if you change the colors:

```
    <Border Background="Blue" BorderBrush="#80ff0000" BorderThickness="20">
```

Gives:

image

You can see the blue background is being drawn under half of the border. I think the simple code should look like:

```
var borderThickness = borders.Left;
var cornerRadius = (float)radii.TopLeft;
var rect = new Rect(size);

if (background != null)
{
	context.FillRectangle(background, rect.Deflate(borders), cornerRadius);
}

if (borderBrush != null && borderThickness > 0)
{
	var strokeOffset = new Thickness(borders.Left / 2, borders.Top / 2, borders.Right / 2, borders.Bottom / 2);
	context.DrawRectangle(new Pen(borderBrush, borderThickness), rect.Deflate(strokeOffset), cornerRadius);
}
```

You can see that's deflating the fill by the border thickness, but deflating the stroke by **half** the border thickness. This is because the rectangle for the stroke designates the _center_ of the stroke:

image

**Comment by grokys** [comment]
However, I must admit that the `rect.Deflate()` method is very confusing, as `rect.Deflate(10)` is not the same as `rect.Deflate(new Thickness(10))`.

**Comment by Gillibald** [comment]
Ahh okay good catch. Haven't checked that. Was afraid there is more involved in a fix. I will have a look at render test next.

**Comment by Gillibald** [comment]
I will close this PR for now and open a new one later this week just to have everything clean and easy to review.
